openedx_learning/apps/authoring/containers/models.py

```diff
@@ -0,0 +1,154 @@
+    )
+    published_version = models.ForeignKey(
+        PublishableEntityVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+        related_name="published_version",
+    )
+
+
+class ContainerEntity(PublishableEntityMixin):
+    """
+    NOTE: We're going to want to eventually have some association between the
+    PublishLog and Containers that were affected in a publish because their
+    child elements were published.
+    """
+
+    pass
+
+
+class ContainerEntityVersion(PublishableEntityVersionMixin):
+    """
+    A version of a ContainerEntity.
+
+    By convention, we would only want to create new versions when the Container
+    itself changes, and not when the Container's child elements change. For
+    example:
+
+    * Something was added to the Container.
+    * We re-ordered the rows in the container.
+    * Something was removed to the container.
+    * The Container's metadata changed, e.g. the title.
+    * We pin to different versions of the Container.
+
+    The last looks a bit odd, but it's because *how we've defined the Unit* has
+    changed if we decide to explicitly pin a set of versions for the children,
+    and then later change our minds and move to a different set. It also just
+    makes things easier to reason about if we say that defined_list never
+    changes for a given ContainerEntityVersion.
+    """
+
+    container = models.ForeignKey(
+        ContainerEntity,
+        on_delete=models.CASCADE,
+        related_name="versions",
+    )
+
+    # This is the EntityList that the author defines. This should never change,
+    # even if the things it references get soft-deleted (because we'll need to
+    # maintain it for reverts).
+    defined_list = models.ForeignKey(
+        EntityList,
+        on_delete=models.RESTRICT,
+        null=False,
+        related_name="defined_list",
+    )
+
+    # inital_list is an EntityList where all the versions are pinned, to show
+    # what the exact versions of the children were at the time that the
+    # Container was created. We could technically derive this, but it would be
+    # awkward to query.
+    #
+    # If the Container was defined so that all references were pinned, then this
+    # can point to the exact same EntityList as defined_list.
+    initial_list = models.ForeignKey(
+        EntityList,
+        on_delete=models.RESTRICT,
+        null=False,
+        related_name="initial_list",
+    )
+
+    # This is the EntityList that's created when the next ContainerEntityVersion
+    # is created. All references in this list should be pinned, and it serves as
+    # "the last state the children were in for this version of the Container".
+    # If defined_list has only pinned references, this should point to the same
+    # EntityList as defined_list and initial_list.
+    #
+    # This value is mutable if and only if there are unpinned references in
+    # defined_list. In that case, frozen_list should start as None, and be
```

Comment:
> This value is mutable if and only if there are unpinned references in `defined_list`. 

In what circumstances do we ever have _pinned_ references in `defined_list`? From what I've seen of the UX designs so far, and everything I know about libraries at least, what we really want is a much simpler world where the draft versions of things in containers are always "unpinned". (And perhaps where the published versions are always pinned, but see my other question about whether or not we want that.)